### PR TITLE
Add quantized CELU operator by adding additional parameters to quantized ELU

### PR DESCRIPTION
**Pull request by durumu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39199 Add quantized CELU operator by adding additional parameters to quantized ELU**

Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

### CODE REVIEW

**Comment by vkuzo** [comment]
there are some failing tests (like https://app.circleci.com/pipelines/github/pytorch/pytorch/174752/workflows/fa9a3d4b-602d-45b3-955f-8c32409c2eb5/jobs/5611015/steps), we should figure out what's going on with them before landing.  The failure looks odd to me, do we have a local repro?

File: torch/nn/quantized/functional.py [hunk 1]
```diff
@@ -478,9 +499,9 @@ def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
         torch._C._nn.elu(input, alpha, out=output)
         return output
     elif inplace:
-        return torch._C._nn.elu_(input, alpha)
+        return torch._C._nn.elu_(input, alpha, 1., 1.)
```
Comment (vkuzo) [inline torch/nn/quantized/functional.py:502]: these have a default of 1 in `native_functions.yaml`, would it work to revert this and next line?

File: aten/src/ATen/native/quantized/cpu/qelu.cpp [hunk 2]
```diff
@@ -29,8 +29,18 @@ Tensor quantized_elu(
     const Tensor& qx, Scalar alpha, Scalar scale, Scalar input_scale) {
   Tensor qy = at::_empty_affine_quantized(qx.sizes(), qx.options(),
       qx.q_scale(), qx.q_zero_point());
-  qelu_stub(qx.device().type(), qx, alpha, qy);
+  qelu_stub(qx.device().type(), qx, alpha, scale, input_scale, qy);
   return qy;
 }
 
+Tensor& quantized_celu_(Tensor& self, Scalar alpha) {
+  double inv_alpha = 1. / alpha.to<double>();
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qelu.cpp:37]: can we verify that division by zero if alpha == 0 gives a good error message?

**Review by vkuzo — COMMENTED** [review]
thanks for working on this!  in general looks great, we should figure out why the derivative tests are failing before landing

File: torch/nn/quantized/functional.py [hunk 3]
```diff
@@ -362,6 +362,27 @@ def max_pool2d(input, kernel_size, stride=None, padding=0, dilation=1,
     return torch.nn.functional.max_pool2d(input, kernel_size, stride, padding,
                                           dilation, ceil_mode, return_indices)
 
+def celu(input, alpha=1., inplace=False):
+    # type: (Tensor, Optional[float], Optional[bool]) -> Tensor
+    r"""celu(input, inplace=False) -> Tensor
+
+    Applies the quantized continuously differentiable ELU function element-wise.
```
Comment (supriyar) [inline torch/nn/quantized/functional.py:369]: ELU -> CELU

Comment (durumu) [reply torch/nn/quantized/functional.py:369]: CELU stands for continuously differentiable ELU -- maybe I should capitalize the C to make it clearer this is what I mean?

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -2601,8 +2601,16 @@
 
 - func: celu(Tensor self, Scalar alpha=1.0) -> Tensor
   use_c10_dispatcher: full
+  dispatch:
+    CPU: celu
+    CUDA: celu
```
Comment (supriyar) [inline aten/src/ATen/native/native_functions.yaml:2606]: Let's try removing these two lines for `celu` and `celu_` below

**Comment by supriyar** [comment]
Changes look good! @vkuzo could you take a look as well?

**Comment by vkuzo** [comment]
looks good, can we verify that `alpha == 0` is handled well with a good error message?

**Comment by durumu** [comment]
@vkuzo I added a new error message that's more verbose - "ZeroDivisionError: alpha cannot be 0 in CELU".

**Review by vkuzo — APPROVED** [review]
looks great! thank you for working on this!

**Comment by ailzhang** [comment]
cc: @dlibenzi @JackCaoG  This PR might require pytorch/xla lowering `celu` .

**Comment by dlibenzi** [comment]
> cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .

We chatted about this time ago as general discussion.
Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.

**Comment by ailzhang** [comment]
> > cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .
> 
> We chatted about this time ago as general discussion.
> Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.

cc: @ezyang that we agreed this is the long term direction we want to achieve. But currently there's no way to allow this as ops have to go either device dispatch or typedefault.
I'll sync with @ezyang to see what's a good plan to get this done. 
But before that feature is available, pytorch/xla will have to move with pytorch master branch to lower it :(

**Comment by dlibenzi** [comment]
> > > cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .
> > 
> > 
> > We chatted about this time ago as general discussion.
> > Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.
> 
> cc: @ezyang that we agreed this is the long term direction we want to achieve. But currently there's no way to allow this as ops have to go either device dispatch or typedefault.
> I'll sync with @ezyang to see what's a good plan to get this done.
> But before that feature is available, pytorch/xla will have to move with pytorch master branch to lower it :(

I guess one more entry in `aten_tensor_ops.cpp/h` 😞

**Review by ailzhang — APPROVED** [review]
I have a companion PR for this in pytorch/xla. Please feel free to rebase and land when it's ready.

**Comment by ailzhang** [comment]
Btw a change of tensorpipe submodule is included in this PR which I think is not related.

**Comment by supriyar** [comment]
> Btw a change of tensorpipe submodule is included in this PR which I think is not related.

@durumu could you remove the tensorpipe submodule update in the PR?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 5]
```diff
@@ -793,7 +793,17 @@ void qtanh_kernel(const Tensor& qx, Tensor& qy) {
   });
 }
 
-void qelu_kernel(const Tensor& qx, Scalar alpha, Tensor& qy) {
+void qelu_kernel(
+    const Tensor& qx,
+    Scalar alpha,
+    Scalar scale,
+    Scalar input_scale,
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:800]: Can you try giving default values to these arguments.

File: test/quantization/test_quantized_op.py [hunk 6]
```diff
@@ -325,6 +325,33 @@ def test_qelu(self, X, alpha):
...
+                                       dtype=torch_type)
+
+        # calculate CELU(dqX) and quantize
+        dqX = qX.dequantize()
+        dqY_hat = torch.nn.functional.celu(dqX, alpha)
+        qY_hat = torch.quantize_per_tensor(dqY_hat, scale=output_scale, zero_point=output_zero_point,
+                                           dtype=torch_type)
+
+        # test regular
+        qY = torch.nn.quantized.functional.celu(qX, output_scale, output_zero_point, alpha=alpha)
```
Comment (supriyar) [inline test/quantization/test_quantized_op.py:350]: I believe you can call `torch.ops.quantized.celu` here instead of functional

**Pull request closed without merging**